### PR TITLE
Vault-managed Azure OpenAI credentials for mit-learn, learn-ai, and edxapp

### PR DESCRIPTION
**Pull request by blarghmatey**
### What are the relevant tickets?

N/A. The design and its rationale live in `docs/plans/vault-azure-openai-credentials-spec.md`, which this PR updates as part of the change.

### Description (What does it do?)

Gives mit-learn, learn-ai and edxapp/mitxonline access to Azure OpenAI using short-lived credentials that Vault mints per lease, instead of a static API key. This is **purely additive** — every app's existing `OPENAI_API_KEY` wiring is untouched, and both providers end up available side by side.

Five pieces:

1. **`OLVaultAzureSecretsEngine` component** (`src/ol_infrastructure/components/services/vault.py`) — mirrors `OLVaultAWSSecretsEngine` next door, but two details differ and are easy to get wrong by copying it: role TTLs are Go duration strings (`"24h"`) rather than integer seconds, and the role name is passed as `role=` rather than `name=`. The second fails *silently*, creating a role under a generated name.

2. **`pulumi-azure-native` + `pulumi-azuread` dependencies.** Worth noting for image size: `pulumi_azure_native` is 155 MB on disk, against the 116 MB `pulumi_aws` we already carry — same order of magnitude as an existing dependency, so no new class of problem, but the venv grows ~160 MB.

3. **`infrastructure/azure/openai`** (new Pulumi project) — per environment, one resource group holding **one Cognitive Services account per app consumer**, three model deployments on each (`gpt-4o`, `gpt-4o-mini`, `gpt-5.2`), and the Azure AD application Vault uses as its own root credential, granted `User Access Administrator` scoped to the resource group rather than the subscription.

4. **`substructure/vault/azure`** (new Pulumi project) — mounts the Azure secrets engine at `azure-openai` with one role per consumer. Each role is scoped to *its own* consumer's account, which is what stops one app's dynamic credentials from reaching another app's endpoint.

5. **Per-app wiring** — Vault policy grants, a dynamic secret, and the non-secret endpoint/tenant env vars for each of the three apps.

Plus the delivery wiring: project-registry constants, a new `pulumi-azure` Concourse pipeline, and `"azure"` added to the Vault substructure chain.

#### Why one account per consumer rather than one shared account

Two Azure behaviours drove this, and both cut against the intuitive design:

- Azure Monitor's Cognitive Services metrics are dimensioned by model deployment name, model name and status code — there is **no stable caller dimension**. Since Vault mints a fresh service principal every lease, the calling identity's object ID churns every 24h, so attributing usage per app by caller cannot work. Attribution has to come from the resource.
- Azure OpenAI TPM quota is allocated per subscription, per region, per model and is **shared across accounts**. More accounts do not buy more throughput; they divide the same pool. Throughput isolation comes from per-deployment capacity allocation, which works in any layout.

One Vault root service principal per environment (scoped to that environment's resource group) keeps this at three admin-consent grants rather than nine.

#### Two findings that changed the spec

Answering the open question about edxapp setting names meant reading the plugin source, which turned up two things the spec had wrong. Both are now corrected in `docs/plans/vault-azure-openai-credentials-spec.md`:

- The spec claimed config-source slot `17-` was free. It is already `17-webhook-tokens-secrets.yaml`, so the Azure file is **`18-`**.
- More significantly: **delivering the settings does not make edxapp able to use them.** The translations plugin reads provider credentials from exactly one place — `settings.TRANSLATIONS_PROVIDERS[provider]["api_key"]` ([`course_translations.py#L64-L76`](https://github.com/mitodl/edx-extensions/blob/main/src/ol_openedx_course_translations/ol_openedx_course_translations/utils/course_translations.py#L64-L76)) — and dispatches to three hardcoded classes. There is no flat `OPENAI_*` read path and no Azure provider. Making this work needs a follow-up in `mitodl/edx-extensions`: an `AzureOpenAIProvider` on LiteLLM's `azure/{deployment}` form, and — the real work — exchanging the `client_id`/`client_secret` for a bearer token via `ClientSecretCredential`, since Vault issues a credential pair rather than a token.

That is why the Azure credentials are delivered as flat top-level settings in their own file rather than as another `TRANSLATIONS_PROVIDERS` block. The edxapp init container concatenates its config sources with `cat` rather than deep-merging them, so a second file emitting that key would produce a duplicate and silently clobber the deepl/openai/gemini/mistral providers — last one wins, no error anywhere. The plugin folds the flat settings into `TRANSLATIONS_PROVIDERS` in Python, where merging actually works.

### How can this be tested?

**Nothing in this PR takes effect on merge.** All three apps' Azure wiring is gated on an `azure_openai_tenant_id` config value that is unset in every stack, so the only change any existing environment sees is an additive Vault policy grant. The gate exists because a `VaultDynamicSecret` aimed at a mount Vault does not have yet *fails* rather than degrading, so the cutover has to be per-environment and after the mount is deployed there.

Verified with real `pulumi preview` runs. In each case the only diff attributable to this change is the Vault policy:

| Stack | Result |
|---|---|
| `ol-application-mit-learn` / `CI` | `~ vault:index:Policy ol-mitlearn-vault-policy-ci` only |
| `ol-application-learn-ai` / `CI` | `~ vault:index/policy:Policy learn-ai-vault-policy-ci` only |
| `ol-application-edxapp` / `mitxonline.CI` | `~ vault:index/policy:Policy edxapp-vault-policy` only |

No resource replaced, no Azure resource created, `OPENAI_API_KEY` untouched, and `TRANSLATIONS_PROVIDERS` / `14-translations-providers-secrets.yaml` unchanged. (Remaining diffs in those previews came from the placeholder Docker tag/digest env vars needed to run a preview locally, plus pre-existing Fastly gzip list-ordering noise.)

Also verified:

- `pre-commit run --all-files` passes. The one failing hook, `hadolint`, fails on pre-existing Dockerfiles that this PR does not touch.
- Both new Concourse pipeline modules generate their `definition.json` without error.
- `tests/ol_concourse/test_secrets_map.py` — all 172 tests pass. This one matters: that suite re-derives the secrets map by static analysis and fails on drift, so the new `secrets_map.py` entries are required rather than optional.
- Every `pulumi-azure-native` / `pulumi-azuread` API shape used here was checked against the installed packages rather than assumed (`Account.properties.endpoint`, `ApplicationPassword.value`, `ServicePrincipal.object_id`, `DeploymentModelVersionUpgradeOption.NO_AUTO_UPGRADE`).

**What has *not* been tested:** the two new Pulumi projects have never been previewed against a real subscription, because no Azure subscription has been chosen yet and the SOPS `azure.{ci,qa,production}.yaml` credential files therefore do not exist.

### Additional Context

**This is reviewable now but not deployable yet.** One open question gates the whole Azure side — which subscription and tenant, whether `Microsoft.CognitiveServices` is registered and approved for Azure OpenAI, and how much regional TPM quota exists. That last one needs a number rather than a yes: this layout puts 27 deployments (9 accounts × 3 models) on one shared regional pool. Current capacity settings are 50k TPM for Production and 5k elsewhere; if the pool is small, that split needs revisiting before the first deploy.

A few decisions worth a reviewer's attention:

- **Model versions are deliberately not pinned in code.** Which versions exist is a property of the subscription and region, not of this repo, and a wrong version string fails at `Deployment` *create* time rather than at preview. Left unset, Azure deploys its current default and upgrades it over time. The project README documents confirming them with `az cognitiveservices account list-models` and then pinning via `azure_openai:model_versions` in stack config — which also flips that deployment to `NoAutoUpgrade`, so a new default version cannot change model behaviour with no deploy and no diff.
- **The apps derive their endpoint rather than reading it from a stack reference.** The infrastructure project sets each account's `custom_sub_domain_name` to the account's own name, so the URL follows from the environment. A `StackReference` would have coupled every mit-learn / learn-ai / edxapp deploy to a stack that does not exist yet — an unacceptable cost for an additive change.
- **The tenant ID doubles as the feature switch** because it is the one value that cannot be derived.
- **`disable_local_auth` is left `false`** on the Cognitive Services accounts. Flipping it is the *final* step of the migration, once nothing depends on key auth.
- **The policies extend `sys/leases/renew` and `revoke` `allowed_parameters`.** Without that the vault-secrets-operator can read a credential once but cannot manage its lease afterwards. (edxapp's policy needed no such change — it grants `sys/leases/renew` without `allowed_parameters`.)
- **`meta.py` and `secrets_map.py` registration** were not in the original plan but are both required. Without the `meta.py` entry the new pipeline file would exist and never be set; without the `secrets_map.py` entries the drift test fails.

### Checklist:
- [ ] Choose the Azure subscription/tenant and confirm `Microsoft.CognitiveServices` registration, Azure OpenAI approval, and regional TPM quota
- [ ] Create the Pulumi deploy service principal out of band and add SOPS `src/bridge/secrets/pulumi/azure.{ci,qa,production}.yaml`
- [ ] Deploy `infrastructure/azure/openai` before `substructure/vault/azure` for each environment
- [ ] Run `az ad app permission admin-consent` once per environment — every resource applies cleanly without it and the failure only surfaces later, when Vault first tries to mint credentials
- [ ] Set `azure_openai_tenant_id` per app, per environment, to switch the app wiring on

### CODE REVIEW

File: src/ol_infrastructure/applications/mit_learn/k8s_secrets.py [hunk 1]
```diff
@@ -339,7 +339,32 @@ def create_mitlearn_k8s_secrets(
...
+    # 7. Dynamic Azure AD credentials for Azure OpenAI from the 'azure-openai' backend.
+    # Vault mints a service principal scoped to mit-learn's own Cognitive Services
+    # account. Additive: OPENAI_API_KEY is untouched and both stay available.
+    #
+    # Gated on mitlearn:azure_openai_tenant_id so this deploys as a no-op until the
+    # substructure/vault/azure mount exists in the environment. Reading a role from a
+    # mount Vault does not have fails the VaultDynamicSecret, so the switch has to be
+    # per-environment rather than at merge time.
+    if Config("mitlearn").get("azure_openai_tenant_id"):
+        azure_openai_secret_name, azure_openai_secret = _create_dynamic_secret(
```
[the remark this answers is not in the document]
Comment (blarghmatey) [reply src/ol_infrastructure/applications/mit_learn/k8s_secrets.py:374]: Good catch, fixed in 15a82bd6d. `_create_dynamic_secret` had no way to express restart targets at all, so it gained an optional `restart_targets` parameter and the Azure secret now passes all five consumers.

On scope: `env_from_secret_names` reaches the celery workers and beat as well as the webapp (k8s.py:1110/1928/2084), so restarting only `mitlearn-app` would have left the workers holding a revoked credential. Targets are `mitlearn-app`, `mitlearn-default-celery-worker`, `mitlearn-edx-content-celery-worker`, `mitlearn-embeddings-celery-worker`, `mitlearn-celery-beat` — taken from `kubectl get deployments -n mitlearn` rather than derived from the naming helper, since a name that doesn't match an existing Deployment fails silently.

File: src/ol_infrastructure/applications/edxapp/k8s_secrets.py [hunk 2]
```diff
@@ -413,6 +416,51 @@ def create_k8s_secrets(
...
+                    "18-azure-openai-secrets.yaml": textwrap.dedent(f"""
+                        AZURE_OPENAI_CLIENT_ID: {{{{ get .Secrets "client_id" }}}}
+                        AZURE_OPENAI_CLIENT_SECRET: {{{{ get .Secrets "client_secret" }}}}
+                        AZURE_OPENAI_TENANT_ID: {azure_openai_tenant_id}
+                        AZURE_OPENAI_ENDPOINT: {azure_openai_endpoint}
+                        AZURE_OPENAI_API_VERSION: {edxapp_config.get("azure_openai_api_version") or "2024-10-21"}
+                        AZURE_OPENAI_DEFAULT_DEPLOYMENT: {edxapp_config.get("azure_openai_default_deployment") or "gpt-5.2"}
+                    """),
+                },
+                vaultauth=vault_k8s_resources.auth_name,
```
[the remark this answers is not in the document]
Comment (blarghmatey) [reply src/ol_infrastructure/applications/edxapp/k8s_secrets.py:463]: Agreed, fixed in 15a82bd6d — the Azure secret now passes `restart_targets=_db_restart_targets`, exactly as you suggested.

Worth noting this one is the most severe of the three: edxapp reads these out of a generated settings file at process start, so without a rollout the LMS and CMS would keep using the expired lease's credentials indefinitely rather than merely until the next pod restart.

File: src/ol_infrastructure/applications/learn_ai/__main__.py [hunk 3]
```diff
@@ -710,8 +710,62 @@
...
+        OLVaultK8SDynamicSecretConfig(
+            name="learn-ai-azure-openai-creds",
+            namespace=learn_ai_namespace,
+            labels=k8s_global_labels,
+            dest_secret_name=azure_openai_secret_name,
+            dest_secret_labels=k8s_global_labels,
+            mount="azure-openai",
+            path="creds/ol-learn-ai-openai",
+            restart_target_kind="Deployment",
+            restart_target_name="learn-ai-app",
```
[the remark this answers is not in the document]
Comment (blarghmatey) [reply src/ol_infrastructure/applications/learn_ai/__main__.py]: Correct, and this was the sharpest of the three catches — fixed in 15a82bd6d. Confirmed the mechanism before changing it: `env_from_secret_names` is applied to the app deployment, the celery worker deployments, and the beat deployment (k8s.py:1110, :1928, :2084), so the single `learn-ai-app` target really would have left the workers and beat on a revoked credential.

Swapped the deprecated `restart_target_kind`/`restart_target_name` pair for the `restart_targets` list covering `learn-ai-app`, `learn-ai-default-celery-worker`, `learn-ai-edx-content-celery-worker` and `learn-ai-celery-beat`. Names verified against `kubectl get deployments -n learn-ai` — the generated worker name replaces the underscore in the `edx_content` queue with a dash, so deriving it by hand would have been easy to get wrong.

File: src/ol_infrastructure/components/services/vault.py [hunk 4]
```diff
@@ -291,6 +291,92 @@ def __init__(
...
+    @field_validator("vault_backend_path")
+    @classmethod
+    def is_valid_path(cls, vault_backend_path: str) -> str:
+        if vault_backend_path.startswith("/") or vault_backend_path.endswith("/"):
+            msg = f"The specified path value {vault_backend_path} can not start or end with a slash"  # noqa: E501
+            raise ValueError(msg)
+        return vault_backend_path
+
+
+class OLVaultAzureSecretsEngine(ComponentResource):
```
[the remark this answers is not in the document]
Comment (blarghmatey) [reply src/ol_infrastructure/components/services/vault.py:331]: Fair — the requirement at `src/ol_infrastructure/AGENTS.md:41-48` is real and this component had no tests. Added `tests/ol_infrastructure/components/services/test_vault_azure_secrets_engine.py` in 15a82bd6d, 10 tests.

They assert what you asked for: that the role is named via the `role` argument, that `ttl`/`max_ttl` render as duration strings rather than integer seconds, that each role's `azure_roles` carries its own `role_name` and `scope`, and that mount-level lease TTLs really are ints (the asymmetry that makes the role-level ones easy to get wrong). There's also a per-role scope test, since that scoping is the isolation boundary between consumers.

I checked the tests actually bite rather than merely passing: reintroducing the `role=` → `name=` trap in the component fails 6 of the 10, including `test_role_is_named_via_role_argument_not_name`.

File: src/ol_concourse/pipelines/infrastructure/azure/pipeline.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    name=Identifier("ol-infrastructure-pulumi-azure"),
+    uri="https://github.com/mitodl/ol-infrastructure",
+    paths=[
+        *PULUMI_WATCHED_PATHS,
+        "src/ol_infrastructure/infrastructure/azure/",
+        *project_secrets_paths("infrastructure/azure/openai/"),
+    ],
+)
+
+azure_openai_fragment = pulumi_jobs_chain(
```
[the remark this answers is not in the document]
Comment (blarghmatey) [reply src/ol_concourse/pipelines/infrastructure/azure/pipeline.py:28]: Verified the mechanism — you're right on the facts. `pulumi_jobs_chain` sets `trigger=True` on the code resource for any non-production stack (ol_concourse `lib/jobs/infrastructure.py:526-528`), and this PR touches the watched path, so the CI job fires on merge and fails decrypting a `pulumi/azure.ci.yaml` that doesn't exist yet.

Raised it with the repo owner rather than deciding unilaterally, and the call is to keep the registration in place. The reasoning: the delivery wiring being omitted is precisely the gap this PR exists to close (see spec §6), and splitting it into a follow-up trades a visible red job for an invisible missing one. The failure is loud, self-explanatory from the log, and clears the moment the subscription question is answered — which is tracked as an explicit pre-merge checklist item on the PR. Leaving as-is deliberately.

File: src/ol_concourse/pipelines/infrastructure/vault/pipeline.py [hunk 6]
```diff
@@ -27,6 +27,11 @@
     "encryption_mounts",
     "secrets",
     "setup",
+    # Depends on the ol-infrastructure-azure-openai stack for the account IDs and the
+    # root service principal it configures the engine with. That project is deployed
+    # by its own pipeline (src/ol_concourse/pipelines/infrastructure/azure/), so this
+    # chain will fail until that one has run for the same environment.
+    "azure",
```
[the remark this answers is not in the document]
Comment (blarghmatey) [reply src/ol_concourse/pipelines/infrastructure/vault/pipeline.py:39]: Confirmed, including the detail that the shared source watches all of `substructure/vault/` so this PR triggers the chain immediately — the comment I wrote acknowledging the dependency was documenting exactly the failure you're describing.

Same decision as the sibling thread on the azure pipeline: taken to the repo owner, and the call is to keep the entry. The azure chain is a separate `pulumi_jobs_chain` fragment, so its failure doesn't block the pki/secrets/auth/setup chains in the same pipeline — it's a visibly red job rather than a blocked deploy path. Weighed against the risk of the substructure never getting wired up at all, that's the accepted tradeoff here. The prerequisite ordering is on the PR's pre-merge checklist.

**Comment by blarghmatey** [comment]
Addressed all six review threads from Copilot — four fixed in 15a82bd6d, two answered with a deliberate decision. All resolved; check bar is green.

**Fixed:** the three rollout-restart findings were correct and worth catching. Vault revokes the service principal at the role's max TTL, but pods read the client id and secret into their environment once at start — so without restart targets Azure auth would have started failing roughly 48h after a deploy rather than at deploy time, which is about the worst shape a failure can take. The reviewer was also right that the scope was wider than the webapp: `env_from_secret_names` reaches the celery workers and beat too, so learn-ai's single `learn-ai-app` target would have left three workloads on a revoked credential. Deployment names were taken from the live `mitlearn` and `learn-ai` namespaces rather than derived, since a target that doesn't match an existing Deployment fails silently. mit-learn's `_create_dynamic_secret` had no way to express this at all and gained an optional parameter.

**Also fixed:** added the Pulumi-mock tests `components/` requires (10 tests), asserting the two API details this component exists to protect — `role=` rather than `name=`, and duration strings rather than integer seconds. I confirmed they actually bite: reintroducing the `role=`/`name=` trap fails 6 of the 10.

**Decided, not fixed:** the two findings about the new pipelines creating guaranteed-failing CI jobs are factually correct — I verified `pulumi_jobs_chain` sets `trigger=True` for non-production stacks, and this PR touches both watched paths. Keeping both registrations anyway, as a deliberate call: missing delivery wiring is the exact gap this PR exists to close (spec §6), and deferring it trades a visible red job for an invisible missing one. The azure substructure chain is a separate fragment, so it doesn't block the other Vault chains. Prerequisite ordering is on the pre-merge checklist in the description.

Still gated on the one open question — which Azure subscription and tenant, and how much regional TPM quota. Nothing here can deploy until that lands.

File: src/ol_infrastructure/applications/learn_ai/__main__.py [hunk 7]
```diff
@@ -710,8 +711,74 @@
...
+            mount="azure-openai",
+            path="creds/ol-learn-ai-openai",
+            # Every workload below gets this secret through the component-wide
+            # env_from_secret_names, so all of them have to restart on rotation,
+            # not just the webapp. Names match what OLApplicationK8s generates:
+            # "{app}-app" for the webapp, "{app}-{worker_name}-celery-worker"
+            # per worker (underscores become dashes), "{app}-celery-beat".
+            restart_targets=[
+                OLVaultRestartTarget(kind="Deployment", name=name)
+                for name in (
```
Comment (Ardiea) [inline src/ol_infrastructure/applications/learn_ai/__main__.py:746]: 💯

**Review by Ardiea — APPROVED** [review]

File: src/ol_infrastructure/applications/learn_ai/__main__.py [hunk 8]
```diff
@@ -779,6 +846,7 @@
             redis_creds_secret_name,
             static_secrets_name,
             opik_keycloak_secret_name,
+            *([azure_openai_secret_name] if azure_openai_tenant_id else []),
         ],
         application_security_group_id=learn_ai_application_security_group.id,
         # Use the fixed name used in the SecurityGroupPolicy spec
```
[the remark this answers is not in the document]
Comment (blarghmatey) [reply src/ol_infrastructure/applications/learn_ai/__main__.py:852]: Moot as of the workload identity rework: there is no azure_openai_secret any more. The design switched from Vault-minted service principals to workload identity federation, so no Kubernetes Secret carries an Azure credential and there is nothing for the Deployment to race against.

The finding was structurally right, though, and the equivalent path in the rework is wired: edxapp's Azure settings ship as a ConfigMap registered in lms_edxapp_config_sources / cms_edxapp_config_sources, and those dicts' values feed depends_on at k8s_resources.py:1004, :1476, :1667, :1790, :1906 and :1318, :2045. mit-learn and learn-ai get plain env vars plus a kubelet-projected token, so neither has a Pulumi-ordered object at all.

File: src/ol_infrastructure/applications/learn_ai/__main__.py (hunk 7 above)
Comment (blarghmatey) [reply src/ol_infrastructure/applications/learn_ai/__main__.py:746]: Resolving as superseded: the rollout-restart targets this was on are deleted in the workload identity rework. Nothing rotates under federation, so there is no rotation for pods to restart on.

**Comment by blarghmatey** [comment]
Resolved the two remaining threads as superseded rather than fixed, because the design changed underneath them.

**Both findings are moot under the replacement design.** The Vault Azure secrets engine path is dead: the root service principal needs Microsoft Graph `Application.ReadWrite.OwnedBy`, and consenting to a Graph app role requires Privileged Role Administrator. We hold Owner on the subscription and zero Entra directory roles, so the portal consent fails with `Authorization_RequestDenied`. Cloud Application Administrator and Application Administrator are both explicitly excluded from consenting to Graph app roles, so the smaller ask would not have worked either.

The replacement is workload identity federation against user-assigned managed identities. Those are ARM resources rather than directory objects, so creating them, federating them to the EKS cluster OIDC issuer, and assigning `Cognitive Services OpenAI User` are all subscription-scope operations already covered by the Owner grant. Zero Entra directory writes, so no admin consent and no IS&T ticket. Nothing rotates, which removes the credential-rotation restart problem along with the Vault mount, the policy grants, and the VSO dynamic secrets.

Sentry's `depends_on` finding was structurally right and the equivalent path in the rework is wired: edxapp's Azure settings ship as a ConfigMap registered in `lms_edxapp_config_sources` / `cms_edxapp_config_sources`, and those dicts' values feed `depends_on` at `k8s_resources.py:1004`, `:1476`, `:1667`, `:1790`, `:1906` and `:1318`, `:2045`. mit-learn and learn-ai get plain env vars plus a kubelet-projected token, so neither has a Pulumi-ordered Kubernetes object at all.

**The red checks are a real finding that carried over.** `test_versions_map.py` fails because `infrastructure/azure/openai/` has no `PROJECT_VERSIONS` entry and the new pipeline watches the project's secrets without watching its version pins. That was still broken in the rework — pre-commit does not run pytest, so it survived a full pre-commit pass. Fixed on the rework branch; full suite is 1008 passed. The entry is an empty list deliberately: model versions are stack config (`azure_openai:model_versions`), not `versions.py` constants, because which versions exist is a property of the subscription and region rather than of this repo.

The checks here will stay red, since the fix is on a different branch. This PR's disposition is still open.

**Comment by blarghmatey** [comment]
Superseded by #5680 and closing.

That PR implements the same goal with workload identity federation against user-assigned managed identities instead of Vault's Azure secrets engine. The reason for the switch is in the previous comment and in section 9 of `docs/plans/azure-openai-credentials-spec.md`: this design's Vault root service principal needs Microsoft Graph `Application.ReadWrite.OwnedBy`, consenting to a Graph app role requires Privileged Role Administrator, and we hold Owner on the subscription with zero Entra directory roles. It cannot be deployed as written.

The replacement performs zero Entra directory writes, so it needs no admin consent and no IS&T ticket, and nothing rotates. The review feedback on this PR carried over where it still applied — in particular the `test_versions_map.py` failures behind the red checks here are fixed in #5680.
